### ISSUE TITLE
Magisk > 21.4 is difficult to install on LineageOS (18.1) with TWRP (Direct Install not enabled unless you use old zip) 

### ISSUE THREAD

**Issue by ElCoyote27** [topjohnwu/Magisk#4394]
Hi,

First of all, thank you for Magisk. Ever since LOS dropped addons I've been using Magisk to allow TitaniumBackup to keep backing up my apps + data.

Device: lithium (Xiaomi Mi Mix)
Android version: Android 11 x86_64 (LineageOS 18.1)
Magisk version name: v23.0
Magisk version code: 23001

I recently wiped my Xiaomi Mi Mix and tried to re-install Magisk by using the 23.0 apk. (I used to use magisk as far back as 17.3 and it was easy to install: from TWRP, Add lineageos-signed.zip, gapps.zip and Magisk.zip and voila, Magisk would work.

On LOS 18.1, I needed to re-install magisk due to a failed update (I had to wipe System + Cache from TWRP and reinstall from scratch) and I ran into something I had never seen before:

Inside the Magisk 23.0 app, my only choice for the 2nd step was "Select and Path a file".
I extracted boot.img from lineageos's signed zip and attempted to use that method. It didn't work and went into recovery.

-HOWEVER- One thing that worked was to do this:

Boot into TWRP, Install the old Magisk-v21.4.zip file and reboot.
After reboot, when going into Magisk 23.0 (the app), I was able to see both choices:
- Select and patch file.
- Direct install (recommended) - that choice wasn't there before.
Using Direct install, I was then able to update to 23.0 without doing the manual patching of a boot.img.

Is that a regression that 22.x and 23.0 are so much more difficult to install than versions up to and including 21.4?

Am I doing something wrong?

**Comment by cdglitch** [comment]
Just updated lavender to 18.1 and can confirm, magisk >21.4 claims I have no boot image and doesn't recognise the twrp recovery image.

Got my old 21.4 zip, flashed from twrp and it works fine. After installing this, update then shows a direct install option and now everything seems to be great. So I guess this is the new required workaround?

**Comment by ElCoyote27** [comment]
@cdglitch Thank you for confirming, I was thinking I was doing something wrong.

**Comment by Tremolo4** [comment]
You can still use the zip install method with a custom recovery. See [here](https://github.com/topjohnwu/Magisk/blob/master/docs/install.md#custom-recovery) in the docs. The trick is to rename the `.apk` to `.zip`. Note the warnings in the docs though: This method might stop working in the future and might already not work on some devices or circumstances.

**Comment by pndwal** [comment]
> Hi,
> 
> First of all, thank you for Magisk. Ever since LOS dropped addons I've been using Magisk to allow TitaniumBackup to keep backing up my apps + data.
> 
> Device: lithium (Xiaomi Mi Mix)
> Android version: Android 11 x86_64 (LineageOS 18.1)
> Magisk version name: v23.0
> Magisk version code: 23001
> 
> I recently wiped my Xiaomi Mi Mix and tried to re-install Magisk by using the 23.0 apk. (I used to use magisk as far back as 17.3 and it was easy to install: from TWRP, Add lineageos-signed.zip, gapps.zip and Magisk.zip and voila, Magisk would work.
> 
> On LOS 18.1, I needed to re-install magisk due to a failed update (I had to wipe System + Cache from TWRP and reinstall from scratch) and I ran into something I had never seen before:
> 
> Inside the Magisk 23.0 app, my only choice for the 2nd step was "Select and Path a file".
> I extracted boot.img from lineageos's signed zip and attempted to use that method. It didn't work and went into recovery.
> 
> -HOWEVER- One thing that worked was to do this:
> 
> Boot into TWRP, Install the old Magisk-v21.4.zip file and reboot.
> After reboot, when going into Magisk 23.0 (the app), I was able to see both choices:
> 
> * Select and patch file.
> * Direct install (recommended) - that choice wasn't there before.
>   Using Direct install, I was then able to update to 23.0 without doing the manual patching of a boot.img.
> 
> Is that a regression that 22.x and 23.0 are so much more difficult to install than versions up to and including 21.4?
> 
> Am I doing something wrong?

You may have lost root initially due to this:
https://github.com/topjohnwu/Magisk/issues/4319

Next, Direct Install is lost as it is root dependant (can be used to update, downgrade or change from public to debug builds, etc).

Initial Install always requires Patch / Flash Image, or Zip Installer using Custom Recovery. PW

**Comment by cdglitch** [comment]
@pndwal I had a very similar experience to ElCoyote but had root throughout my attempts to get magisk up and running. At no point before flashing my old magisk .zip was I offered the direct install method.

The bug IMO is that the magisk app does not detect the boot image on LineageOS 18.1 but while flashing the old .zip it clearly says it finds the boot image and operates on it normally. Post install the app still does not recognize the presence of a boot image, but it does find a magisk installation that it can upgrade via direct install. Which it then does with no issues and patches the boot image.

**Comment by pndwal** [comment]
> @pndwal I had a very similar experience to ElCoyote but had root throughout my attempts to get magisk up and running. At no point before flashing my old magisk .zip was I offered the direct install method.
> 
> The bug IMO is that the magisk app does not detect the boot image on LineageOS 18.1 but while flashing the old .zip it clearly says it finds the boot image and operates on it normally. Post install the app still does not recognize the presence of a boot image, but it does find a magisk installation that it can upgrade via direct install. Which it then does with no issues and patches the boot image.

Ok, looks like there's more to it.

Still, I'm confused by 'magisk >21.4 claims I have no boot image and doesn't recognise the twrp recovery image.'

Please can you clarify what 'magisk >21.4 claims ...' and what 'doesn't recognise the twrp recovery image' mean? ... I'm guessing you may mean 'claims no ramdisk' and 'patched Recovery image doesn't work'. ??
PW

**Comment by ElCoyote27** [comment]
@pndwal I'm going to try to explain what I saw:

I have a Xiaomi Mi Mix phone (6G RAM, 256G storage) with TWRP. I run LineageOS on this one (since 14.1) and I had been using Magisk to get root on LOS (LineageOS) since Magisk 17.x or Magisk 18.x.
Things had been working well and I was able to upgrade from LOS 16.0 to 17.1 to 18.1 over the course of a few years.

A few days ago a LOS 18.1 -update- failed and I had to wipe 'System' and 'Cache' and reinstall LOS from scratch.
Since I have a working TWRP, it's easy to do:
- Boot into recovery, advanced wipe system + cache, install zips (lineageos-signed-20210513.zip, + mindthegapps + Magisk v22.1 zip file), as I always did.

Right after reboot, I noticed that my phone was no longer rooted. When I launched the Magisk App (I had 23.0 installed), the only choice to reinstall/update magisk was to 'select a file to patch'. I tried it and it didn't work for me.

At the end of a long day of investigation, I stumbled onto a workaround:
1) boot into TWRP and install the -old- 21.4 zip file. This one:
````# ls -la Magisk-v21.4.zip 
-rw-r--r--. 1 root root 6173806 Jan 17 09:16 Magisk-v21.4.zip
# sha256sum Magisk-v21.4.zip
35369e23bc565888e8d8ec62294edaaae98ef4e15ba81349274cd74f8eb3c214  Magisk-v21.4.zip
```

2) reboot into LOS (the phone was now rooted but Magisk detected that I had 21.4 installed).
When going into the Magisk App, the 'Direct Install (recommended)' otion was now enabled and I was able to update Magisk to 23.0 (after reboot it did show 23.0 as installed).

In short, those two zip files didn't work for me on TWRP + LOS:

```# ls -la Magisk-2*
-rw-r--r--. 1 root root 17660819 May 28 13:03 Magisk-22.1.zip
-rw-r--r--. 1 root root 17609878 May 28 13:03 Magisk-23.0.zip
# sha256sum Magisk-2*
9fc78bf77edba52520e3aa5ab08fbe2767d113f8a5eee8c21d6d3a0dcbb5e67c  Magisk-22.1.zip
1a9fa043ba7abe5fdd8d9308b4d9bfe2216b0325b1d97f935148d10861b68158  Magisk-23.0.zip
```

I had to downgrade to 21.4 first and then upgrade back to 23.0
I think something is missing from the new 17Mb+ zip files you guys have been producing after 21.4.

**Comment by ElCoyote27** [comment]
@pndwal Do you need further clarification? I'm happy to provide additional details if needed.

**Comment by pndwal** [comment]
> @pndwal Do you need further clarification? I'm happy to provide additional details if needed.

Thanks for ping... missed earlier.

Looks like issue (some regression) w/ 22.1 since seems root failed. There were bootloops on many devices w/ LOS 18.1 also w/ 22.1.
https://github.com/topjohnwu/Magisk/issues/4213

... Possibly related to dirty flash / mismatched App etc also....

Of course, there will be no fix for this as newer Magisk are likely free of issue already, so this issue can probably be closed.

Nb. Many are now moving to Alpha Magisk builds (progression of Canary; different dev):
https://forum.xda-developers.com/t/magisk-general-support-discussion.3432382/post-85103975

I'd read recent pages in that thread to get up to speed too. 😉 PW

**Comment by cdglitch** [comment]
> Still, I'm confused by 'magisk >21.4 claims I have no boot image and doesn't recognise the twrp recovery image.'

So the newer magisk versions >21.4 which don't seem to recommend flashing the old magisk .zip from recovery, the app gives me the following; Installed: no, ramdisk: no, A/B: no, SAR: yes. For this version of the app if I want to install magisk it only gives me one option: recovery mode. It then asks me to point it to a recovery.img for it to patch. If I point it to the image of twrp that I use it flat out refuses to patch it as it claims it isn't a valid boot image. This is what I meant for the second point:
 
> Please can you clarify what 'magisk >21.4 claims ...' and what 'doesn't recognise the twrp recovery image' mean? ... I'm guessing you may mean 'claims no ramdisk' and 'patched Recovery image doesn't work'. ??


Once I flash an old .zip for an old magisk, that same app versions then sees that magisk is installed and says I should update it. And when I do this, it gives me a new option for a direct install which isn't there before and everything works fine. 

For me the bug is that it does not give me the option for a direct install if it thinks there is no ramdisk image to boot from. Clearly from how I've gotten Magisk to work there is one of two things happening:
1. a way to get a direct install working without a ramdisk exists and the existing method for updating Magisk is already making it work
2. I actually have a ramdisk and its not being detected and the logic used update/install check gets around the ramdisk not being detected

**Comment by bestouff** [comment]
For people in that case, is there a simple way to get root again, just for doing backups before redoing a clean flash ?

**Comment by ElCoyote27** [comment]
@bestouff Install the 21.4 zip, then use the Magisk App to ugprade to 23.0 using Direct Install. This is what worked for me on LOS 18.1

**Comment by bestouff** [comment]
That did not work for me. My phone has been stuck on the boot animation for at least 1 hour now.

**Comment by ElCoyote27** [comment]
One thing to note: even if you upgrade to 23.0, then it stops working the next time there is a LOS 18.1 update. I had to wipe my system + cache and re-do the 21.4 zip. I won't be upgrading to 23.0 again.

**Comment by bestouff** [comment]
Perhaps, but in my case just flashing 21.4 prevents my phone (OnePlus 7 T Pro) from booting (it used to work with LineageOS 17.1)

**Comment by mrsolus** [comment]
Try another rom

**Comment by pndwal** [comment]
> @pndwal I had a very similar experience to ElCoyote but had root throughout my attempts to get magisk up and running. At no point before flashing my old magisk .zip was I offered the direct install method.
> 
> The bug IMO is that the magisk app does not detect the boot image on LineageOS 18.1 but while flashing the old .zip it clearly says it finds the boot image and operates on it normally. Post install the app still does not recognize the presence of a boot image, but it does find a magisk installation that it can upgrade via direct install. Which it then does with no issues and patches the boot image.

Just revisiting this, and finally clicked re. what you mean by  'magisk >21.4 claims I have no boot image and doesn't recognise the twrp recovery image.' (I think).

Are you saying it says Ramdisk: no, but won't patch Recovery image?

Well Xiaomi devices generally break the mold (exception to the rule)

While Magisk App is correct that they lack FACTORY ramdisk in boot.img (as do all type III devices), they do inexplicably have bootloader support for boot ramdisk, which is just as well since recovery ramdisk generally doesn't work for Magisk on Xiaomi either.

Since John Wu added commit to add basic ramdisk when missing, patching adds it, but App still reports factory state - Ramdisk: no - since Magisk cannot detect bootloader support and therefore an added boot ramdisk cannot be relied upon. However experience shows it works in most Xiaomi devices 

This is proved by the fact custom Recovery zip method works for Xiaomi type III (Ramdisk no) devices; it ALWAYS patches boot.img! ... It generally doesn't work for these devices, and they require recovery image patching w/ Recovery mode selected in Magisk App.

So, with Xiaomi, despite Ramdisk: no, image patching will also work, but Recovery mode must be de- selected, boot.img patched and boot partition flashed.

**Comment by AndroidDeveloperLB** [comment]
Say, where do I get the zipped flashable files of Magisk? All I've found was from non-official websites for this. Weird that I can't find them here...

**Comment by pndwal** [comment]
I think you'd better read Installation Instruction page. Tip: Custom Recovery heading. PW

**Comment by AndroidDeveloperLB** [comment]
@pndwal That's what I did:
https://topjohnwu.github.io/Magisk/install.html#custom-recovery
No zip file there.
Now we can only patch the boot.img file? We can't use a single ZIP file to flash?

**Comment by androidacy** [comment]
Rename APK to zip and flash

**Comment by AndroidDeveloperLB** [comment]
@androidacy This works? 
Anyway I got the "boot.img" from the custom-ROM file, and patched it. Seems to work fine.
Weird that you wrote about renaming though. Didn't think it's possible. Maybe will try next time.

**Comment by pndwal** [comment]
> @pndwal That's what I did:
> https://topjohnwu.github.io/Magisk/install.html#custom-recovery
> No zip file there.

Quote:
"- Download the Magisk APK
- Rename the .apk file extension to .zip, for example: Magisk-v22.0.apk → Magisk-v22.0.zip. If you have trouble renaming the file extension (like on Windows), use a file manager on Android or the one included in TWRP to rename the file.
- Flash the zip just like any other ordinary flashable zip."

> Now we can only patch the boot.img file? We can't use a single ZIP file to flash?

🙄

**Comment by AndroidDeveloperLB** [comment]
@pndwal ok you are right. I should stop asking questions so late at night. Sorry about this. Quite surprising it works though, and cool!

**Comment by pndwal** [comment]
> @pndwal ok you are right. I should stop asking questions so late at night. Sorry about this. Quite surprising it works though, and cool!

Changelog:
v22.0
[General] Magisk and Magisk Manager is now merged into the same package!
https://topjohnwu.github.io/Magisk/releases/22000.html#magisk-manager-is-deadlong-live-the-magisk-app
😉 PW

**Comment by pndwal** [comment]
Sorry if a bit late!

> If I point it to the image of twrp that I use it flat out refuses to patch it as it claims it isn't a valid boot image.

Don't patch recovery. Patch boot.img w/ Recovery mode deselected. Xiaomi breaks the Ramdisk = no mold. See below.

> For me the bug is that it does not give me the option for a direct install if it thinks there is no ramdisk image to boot from.

Not quite. Please see what I said above:

> Direct Install is lost as it is root dependant (can be used to update, downgrade or change from public to debug builds, etc).
Initial Install always requires Patch / Flash Image, or Zip Installer using Custom Recovery. PW

Direct Install (eg to update, downgrade or change to different channel build) is supported only with root.
If you have root, but see Magisk Installed: N/A, you have a conflicting App or Manager previously not restored / unhidden (it's also renamed) before updating Magisk, or possibly a bug / regression. In these cases you won't see Direct Install option either.

> Clearly from how I've gotten Magisk to work there is one of two things happening:
> 1. a way to get a direct install working without a ramdisk exists and the existing method for updating Magisk is already making it work

Yes, for two reasons:
1) You gain root which is requisite.
2) Magisk actually adds basic ramdisk to boot.img where missing since this 2019 commit:
https://github.com/topjohnwu/Magisk/commit/e72c6685edf81706617a3444575c4500b9b8fe6c

> Since magiskinit can run independently on system-as-root devices, we simply just create an empty ramdisk with magiskinit added as init.

So, if missing, ramdisk is now in fact ALWAYS added to boot image when patched by Magisk app. It just won't work unless boot partition ramdisk support (inexplicably in the case of type III Xiaomi) exists in bootloader.

Nb. 'Ramdisk = no' remains despite manually added ramdisk in boot image because this DOES NOT normal work in type III devices. So Magisk simply reports if FACTORY ramdisk is present as this is the only one that guarantees bootloader support and bootloader CANNOT be polled. (Factory OEM ramdisk indicates compatible bootloader. Manually added ramdisk indicates nothing!) 
See last paragraph of this doc:
https://topjohnwu.github.io/Magisk/boot.html

> 2. I actually have a ramdisk and its not being detected and the logic used update/install check gets around the ramdisk not being detected

No, ramdisk is absent until Magisk adds (as explained above). Even after that, Magisk still indicates Factory ramdisk state = no for reasons explained, but this in no impedes patching process as long as Recovery mode is NOT selected.

Aside from unchecking Recovery mode, be sure to Unhide (restore) App if hidden as this has caused issues with patching process (strangely). PW

**Comment by madduck** [comment]
I tried the 21.4 trick after sideloading LOS 18.1 onto my OnePlus 5T (dumpling), but it seems to just enter the same boot loop as it entered after I installed the patched 23.0 `boot.img` using the official instructions. Note that I verified that LOS 18.1 otherwise boots just fine.

Everything seems to be going fine until suddenly, there are a couple of failures, resulting in the system to reboot into recovery:

```
06-30 13:45:13.849   601   601 E SELinux : avc:  denied  { add } for pid=20375 uid=1000 name=packagename.service scontext=u:r:system_server:s0 tcontext=u:object_r:default_android_service:s0 tclass=service_manager permissive=0
06-30 13:45:13.850 20375 20375 I EdXposed-Bridge: FakeGApps: Adding the package name service failed.
06-30 13:45:13.850 20375 20375 I SystemServerTiming: InitWatchdog
06-30 13:45:13.850 20375 20375 I SystemServerTiming: StartOverlayManagerService
06-30 13:45:13.859     0     0 I [20210630_13:45:13.859114]@4 init: Control message: Processed ctl.start for 'idmap2d' from pid: 20375 (system_server)
06-30 13:45:13.865  1064  1841 W cutils  : Expected path /data/user_de/0/com.markusborg.test with mode 700 but found 751
06-30 13:45:13.884 20413 20413 I uncrypt :   received command: [--prompt_and_wipe_data
06-30 13:45:13.884 20413 20413 I uncrypt : --reason=RescueParty
06-30 13:45:13.884 20413 20413 I uncrypt : --locale=en_AU
06-30 13:45:13.884 20413 20413 I uncrypt : ] (59)
06-30 13:45:13.885 20413 20413 I uncrypt :   received 0, exiting now
06-30 13:45:13.885 20375 20412 I RecoverySystemService: uncrypt setup bcb successfully finished.
06-30 13:45:13.887     0     0 I [20210630_13:45:13.886542]@5 init: Received sys.powerctl='reboot,recovery' from pid: 20375 (system_server)
```
I am happy to make a full logcat available upon request, don't really want to post it here though. Have now reverted to backup, but am happy to try this again (even using IRC/Matrix live chat) whenever if it helps with debugging.

**Comment by 2shrestha22** [comment]
It is very annoying to re root each time when you do OTA update. There are 2 options flash apk (directly flashing apk without renaming to zip also works) in twrp or patch boot.img from magisk manager. Flashing apk is deprecated and we only have a option which is to path boot.img then flash.

Today I did the following which I feel easier to download extract boot img after every OTA.

- I backed up boot partition using TWRP.
- Then patched boot.emmc.win with magisk and it was successful and created a new patched file .img
- I flashed patched img in boot partition.

I was not sure if it will work or not but worked. So I don't have to download boot.img anymore.

**Comment by bgreggsmith** [comment]
> It is very annoying to re root each time when you do OTA update. There are 2 options flash apk (directly flashing apk without renaming to zip also works) in twrp or patch boot.img from magisk manager. Flashing apk is deprecated and we only have a option which is to path boot.img then flash.

This. I have 2 people on the lineageOS 18.1 build for lavender in addition to myself. Both use magisk for a couple shitty apps that only work on a handful of samsung handsets or require safetynet. As it is, I can't recommend they update to 18.1 because *every* update wipes magisk and you have to reboot again to reflash it in recovery, go back into the manager app after rebooting once more to update it by navigating the check boxes and then rebooting *again*.

At the same time lineage 17.1 is becoming old so if anyone has any insight on how we might fix this to how it was I'm all ears and willing to spend some time testing things etc.

**Comment by christophehenry** [comment]
It also seem like LineageOS decided to stop supporting TWRP. When migrating my OPO 5T from Lineage 17.1 to 18.1, the update wiped TWRP's recovery to install Lineage's without asking me first. I suspected this was why I keep loosing the the root acces after avery update but it appears I'm not the only one not being able to use Magisk's direct install. For some time, I was able to repack Lineage's recovery using magisk and flash it but last time this solution just wouldn't work.

For now, sideload update using APK works. But for how long?

**Comment by bgreggsmith** [comment]
> It also seem like LineageOS decided to stop supporting TWRP. When migrating my OPO 5T from Lineage 17.1 to 18.1, the update wiped TWRP's recovery to install Lineage's without asking me first. I suspected this was why I keep loosing the the root acces after avery update but it appears I'm not the only one not being able to use Magisk's direct install. For some time, I was able to repack Lineage's recovery using magisk and flash it but last time this solution just wouldn't work.
> 
> For now, sideload update using APK works. But for how long?

None of us are on TWRP anymore and this is still an issue. I used to use TWRP but during the 17.1 -> 18.1 update TWRP bricked my phone twice whereas on the lineageos recovery its a pretty smooth process. I'm flashing magisk through their recovery at the moment. Gapps and root do something correctly since they don't get wiped on every update for me but magisk requires manual intervention and 3 reboots for it to work after every update.

**Comment by madduck** [comment]
I still have TWRP with 18.1, and I do not recall doing anything specific to keep it that way. I also do not know how to switch to LineageOSs recovery, as that doesn't seem to be an option even.

@bgreggsmith Would you please be able to provide the steps you take during upgrades, just so we have a working procedure documented here for everyone?

**Comment by osm0sis** [comment]
Sounds more like you should be investigating why addon.d isn't working on your LOS updates.

**Comment by pojavluncher2021** [comment]
Sh means shell I think
So, Either the shell is too old or magisk-protected (Safe Shell)
Also how's oplus.fstab doing?

**Comment by osm0sis** [comment]
This still an issue with 23012?

**Comment by ElCoyote27** [comment]
@osm0sis Where is 23012 to be found? My App only shows 23000 still.

**Comment by osm0sis** [comment]
Install Canary app and switch to Canary in settings.

**Comment by osm0sis** [comment]
The app downloads the apk, just press there update button again. Then, modern TWRP allows Magisk*.apk to be flashed as a zip, so no need to rename.

Glad this got resolved upstream! Closing now. 🙂
